src/lisp/kernel/lsp/iolib.lisp

```diff
@@ -302,10 +302,32 @@ the one defined in the ANSI standard. *print-base* is 10, *print-array* is t,
 	   (si:cons-cdr %progv-list)
 	 ,@body))))
 
-(defmacro print-unreadable-object
-	  ((object stream &key type identity) &body body)
-  (if body
-      `(flet ((.print-unreadable-object-body. () ,@body))
-	 (print-unreadable-object-function
-	   ,object ,stream ,type ,identity #'.print-unreadable-object-body.))
-    `(print-unreadable-object-function ,object ,stream ,type ,identity nil)))
+(defun print-unreadable-object-contents (object stream type identity body)
+  (when type
+    (write (type-of object) :stream stream :circle nil
+                            :level nil :length nil)
+    (write-char #\space stream))
+  (when body
+    (funcall body))
+  (when identity
+    (when (or body (not type))
+      (write-char #\space stream))
+    (write-char #\@ stream)
+    (core:write-addr object stream)))
+
+;;; The guts of print-unreadable-object, inspired by SBCL.
+;;; pprint-logical-block isn't available yet, so this function
+;;; will be redefined later.
```

Comment:
Could you note that specifically it's redefined in pprint.lisp? And possibly indicate in pprint.lisp that it is a redefinition. I try to make things really clear when we have to jump through bootstrapping hoops like this.

Reply:
Of course!